### PR TITLE
[1/n] Add support for iterable datasets to Rich ProgressBar

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Related #9647 

This adds support to rich for iterable datasets. Also included are a few components required to make it work + some improvements to the logic as there were a few visual glitches here.

Also included are changes to support DDP Spawn progress, however these tests are a part of #9647 and will be introduced in that PR.

demo
## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -61,17 +129,16 @@ def render(self, task) -> RenderableType:
...
             self._pl_module = pl_module
-            self._stage = stage
             self._tasks = {}
             self._current_task_id = 0
             super().__init__()
 
         def render(self, task) -> Text:
-            if self._stage != "fit" or self._trainer.sanity_checking:
-                return ""
+            if self._trainer.state.fn != "fit" or self._trainer.sanity_checking:
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/rich_progress.py:140]: can you use TrainerFn here?

**Review by rohitgr7 — APPROVED** [review]
LG! just one small comment

**Pull request merged**